File: java/src/main/java/com/purescraps/purehtml/transformers/types/CapitalizeTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.util.List;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

/**
 * capitalize: Uppercase the first character of a string and lowercase the
 * rest, e.g. "hELLO WORLD" -> "Hello world".
 */
public class CapitalizeTransformer extends Transformer {
    private List<String> args;

    public CapitalizeTransformer(List<String> args) {
        this.args = args;
    }

    public CapitalizeTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    public String getName() {
        return "capitalize";
    }

    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (!(val instanceof String)) {
            throw new IllegalArgumentException("capitalize: expected string. Got " + val);
        }

        String str = (String) val;
        if (str.isEmpty()) {
            return str;
        }

        String lower = str.toLowerCase();
        return Character.toUpperCase(lower.charAt(0)) + lower.substring(1);
    }
}

File: python/purehtml/transformers/types/capitalize_transformer.py
from abc import ABC
from typing import List

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class CapitalizeTransformer(Transformer, ABC):
    """
    capitalize: Uppercase the first character of a string and lowercase the
    rest, e.g. "hELLO WORLD" -> "Hello world".
    """

    def __init__(self, args: List[str] = None):
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        self.args = args

    def get_name(self) -> str:
        return "capitalize"

    def transform(self, params: TransformParams):
        val = params.get_val()

        if val is None:
            return None

        if not isinstance(val, str):
            raise TypeError(f"capitalize: expected string. Got {val!r}")

        return val.capitalize()
